lang/java/reef-runtime-standalone/src/main/java/org/apache/reef/runtime/standalone/driver/NodeListManager.java

```diff
@@ -0,0 +1,304 @@
+    this.fileNames = fileNames;
+    this.jvmHeapFactor = 1.0 - jvmHeapSlack;
+    this.reefEventHandlers = reefEventHandlers;
+    this.errorHandlerRID = remoteManager.getMyIdentifier();
+    this.rootFolder = rootFolder;
+    this.nodeFolder = nodeFolder;
+
+    this.nodeSetIterator = this.nodeInfoSet.iterator();
+
+    remoteManager.registerHandler(ReefServiceProtos.RuntimeErrorProto.class,
+        new EventHandler<RemoteMessage<ReefServiceProtos.RuntimeErrorProto>>() {
+          @Override
+          public void onNext(final RemoteMessage<ReefServiceProtos.RuntimeErrorProto> value) {
+            final FailedRuntime error = new FailedRuntime(value.getMessage());
+            LOG.log(Level.SEVERE, "FailedRuntime: " + error, error.getReason().orElse(null));
+            release(error.getId());
+          }
+        });
+
+    LOG.log(Level.FINEST, "Initialized NodeListManager.");
+  }
+
+  void release(final String containerID) {
+    synchronized (this.containers) {
+      final Container ctr = this.containers.get(containerID);
+      if (null != ctr) {
+        LOG.log(Level.INFO, "Releasing Container with containerId [{0}]", ctr);
+        if (ctr.isRunning()) {
+          ctr.close();
+        }
+        this.containers.remove(ctr.getContainerID());
+      } else {
+        LOG.log(Level.INFO, "Ignoring release request for unknown containerID [{0}]", containerID);
+      }
+    }
+  }
+
+  public void onResourceLaunchRequest(final ResourceLaunchEvent resourceLaunchEvent) {
+    LOG.log(Level.INFO, "NodeListManager:onResourceLaunchRequest");
+
+    final String remoteNode;
+    try {
+      synchronized (this.nodeSetIterator) {
+        remoteNode = this.getNode();
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to get remote node", e);
+    }
+    final String username = remoteNode.substring(0, remoteNode.indexOf('@'));
+    final String hostname = remoteNode.substring(remoteNode.indexOf('@') + 1);
+    final String userHomeDir = System.getProperty("user.home");
+    final String privatekey = userHomeDir + "/.ssh/id_dsa";
+
+    synchronized (this.containers) {
+      final JSch remoteConnection = new JSch();
+      final Session sshSession;
+      try {
+        remoteConnection.addIdentity(privatekey);
+        sshSession = remoteConnection.getSession(username, hostname, sshPort);
+
+        final Properties jschConfig = new Properties();
+        jschConfig.put("StrictHostKeyChecking", "no");
+        sshSession.setConfig(jschConfig);
+
+        try {
+          sshSession.connect();
+        } catch (JSchException ex) {
+          throw new RuntimeException("Unable to connect to the session. Check your authorized_keys settings.", ex);
+        }
+
+        LOG.log(Level.FINEST, "Established connection with {0}", hostname);
+
+        final Container c = this.containers.get(resourceLaunchEvent.getIdentifier());
+        c.setRemoteConnection(sshSession, remoteNode);
```

Comment:
`Container` seems to be doing two things
1) onResourceRequest: Save resource info
2) onLaunchRequest: Save ssh connection/files info

For 1), you are using one-time constructor, making the variables final.
For 2), you are `setting` the values. This could cause troubles such as duplicate setting of values.

Could you please also somehow make 2) immutable?